WordPress/Classes/Utility/iAds/SearchAdsAttribution.swift

```diff
@@ -0,0 +1,123 @@
+/// Implementation of Search Ads attribution details
+/// More info: https://searchads.apple.com/help/measure-results/#attribution-api
+
+import Foundation
+import iAd
+import AutomatticTracks
+
+@objc final class SearchAdsAttribution: NSObject {
+
+    /// Keep the instance alive
+    ///
+    private static var lifeToken: SearchAdsAttribution?
+
+    private static let userDefaultsSentKey = "search_ads_attribution_details_sent"
+    private static let userDefaultsLimitedAdTrackingKey = "search_ads_limited_tracking"
+
+    private let searchAdsApiVersion = "Version3.1"
+
+    /// Is ad tracking limited?
+    /// If the user has limited ad tracking, and this API won't return data
+    ///
+    private var isTrackingLimited: Bool {
+        get {
+            return UserDefaults.standard.bool(forKey: SearchAdsAttribution.userDefaultsLimitedAdTrackingKey)
+        }
+        set {
+            UserDefaults.standard.set(newValue, forKey: SearchAdsAttribution.userDefaultsLimitedAdTrackingKey)
+        }
+    }
+
+    /// Has the attribution details been sent already?
+    /// Once the attribution details are sent, it won't change; so it is not necesary to send it again
+    ///
+    private var isAttributionDetailsSent: Bool {
+        get {
+            return UserDefaults.standard.bool(forKey: SearchAdsAttribution.userDefaultsSentKey)
+        }
+        set {
+            UserDefaults.standard.set(newValue, forKey: SearchAdsAttribution.userDefaultsSentKey)
+        }
+    }
+
+    override init() {
+        super.init()
+        SearchAdsAttribution.lifeToken = self
```

Comment:
Suggestion / Thought: with the current API you could have multiple `SearchAdsAttribution` instances, and just one pointer to the active one.

How about if we:

A. Set the initializer as `private` (not sure how that'd work, since this kid inherits from NSObject)
B. Add a guard so that there can only be one instance. Maybe something like this would suffice:

```
guard lifeToken == nil else {
   fatalError("There can only be one!")
}
```

C. Simpler, maybe, a singleton?. I know we all hate singletons, though!.

Reply:
@jleandroperez  I see your point!

I didn't want it to be singleton to not keep it in memory unnecessarily for the whole lifetime of the app, specially when we need it just once in just one place. But if that's not a problem with this small class, I'd go for the simplicity and make it a singleton.

The traditional way would be that the caller keeps it alive with a property and it is released via completion handler.

What do you think?

Reply:
Personally, i would go with an approach like this:

```
class A {
   private var internal: A?
   private init() { }   /// << not sure if this builds, when A inherits from NSObject
   static var instance: A {
        let output = internal ?? A()
        if internal == nil {
             internal = output
        }
        return output
   }
}
```

That way you get the best of both worlds: the instance can get released whenever you need, and you've got a mechanism that would prevent multiple instances from being around.

Big note is: the snippets i've pasted are *not* thread safe!

Reply:
Great! I will give it a try